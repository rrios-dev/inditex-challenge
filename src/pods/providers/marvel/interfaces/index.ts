export * from "./comic";
export * from "./common";
export * from "./hero";
