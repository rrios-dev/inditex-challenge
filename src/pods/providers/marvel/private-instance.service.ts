import crypto from "crypto";

import axios from "redaxios";

import config from "@/pods/config";
import serverConfig from "@/pods/server-config";

const ts = Date.now();

const privateMarvelInstanceService = axios.create({
  baseURL: config.api.public.url,
  params: {
    ts: Date.now(),
    apikey: config.api.public.key,
    hash: crypto
      .createHash("md5")
      .update(`${ts}${serverConfig.api.private.key}${config.api.public.key}`)
      .digest("hex"),
  },
});

export default privateMarvelInstanceService;
