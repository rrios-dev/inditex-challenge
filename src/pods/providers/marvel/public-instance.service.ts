import axios from "redaxios";

import config from "@/pods/config";

const publicMarvelInstanceService = axios.create({
  baseURL: config().api.public.url,
  params: {
    apikey: config().api.public.key,
  },
});

export default publicMarvelInstanceService;
