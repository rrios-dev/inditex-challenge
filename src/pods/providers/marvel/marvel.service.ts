import axios, { Options } from "redaxios";
import { mergeDeep } from "remeda";

import { Comic, Hero, MarvelPagination, MarvelResponse } from "./interfaces";
import publicMarvelInstanceService from "./public-instance.service";

type CustomOptions = Options & {
  instance?: ReturnType<typeof axios.create>;
};

interface GetHeroListInput {
  search?: string;
  offset?: number;
  limit?: number;
}
export const getHeroList = async (
  input?: GetHeroListInput,
  options: CustomOptions = {}
) =>
  (options?.instance || publicMarvelInstanceService).get<
    MarvelResponse<MarvelPagination<Hero[]>>
  >(
    "characters",
    mergeDeep(options, {
      params: {
        limit: input?.limit ?? 50,
        offset: input?.offset ?? 0,
        ...(input?.search && { nameStartsWith: input.search }),
      },
    })
  );

export const getHeroDetail = async (id: string, options?: CustomOptions) =>
  (options?.instance || publicMarvelInstanceService).get<
    MarvelResponse<MarvelPagination<[Hero]>>
  >(`characters/${id}`, options);

export const getComic = async (id: string, options?: CustomOptions) =>
  (options?.instance || publicMarvelInstanceService).get<
    MarvelResponse<MarvelPagination<[Comic]>>
  >(`comics/${id}`, options);
