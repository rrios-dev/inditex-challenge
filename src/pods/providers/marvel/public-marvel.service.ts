import { Options } from "redaxios";

import { Comic, Hero, MarvelPagination, MarvelResponse } from "./interfaces";
import publicMarvelInstanceService from "./public-instance.service";

export const getCharacters = async (options?: Options) =>
  publicMarvelInstanceService.get<MarvelResponse<MarvelPagination<Hero[]>>>(
    "characters",
    options
  );

export const getCharacter = async (id: string, options?: Options) =>
  publicMarvelInstanceService.get<MarvelResponse<MarvelPagination<[Hero]>>>(
    `characters/${id}`,
    options
  );

export const getComic = async (id: string, options?: Options) =>
  publicMarvelInstanceService.get<MarvelResponse<MarvelPagination<[Comic]>>>(
    `comics/${id}`,
    options
  );
