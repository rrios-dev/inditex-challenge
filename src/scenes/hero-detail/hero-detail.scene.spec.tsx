import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import FavsProvider from "@/pods/favs/favs-provider";
import * as service from "@/pods/providers/marvel/marvel.service";

// import { emptyHeroListMock, heroListMock } from "./__mocks__/get-hero-list";
import { heroDetailMock } from "./__mocks__/get-hero-detail";
import HeroDetail from "./hero-detail.scene";

vi.mock("react", async (importOriginal) => ({
  ...((await importOriginal()) as any),
  cache: () => () => ({}),
}));

vi.mock("next/navigation", async (importOriginal) => ({
  ...((await importOriginal()) as any),
  useParams: () => ({ id: "1234-" }),
}));

describe("Hero Detail Scene", () => {
  it.only("Mount Scene with data", async () => {
    vi.spyOn(service, "getHeroDetail").mockReturnValue(
      Promise.resolve({
        data: heroDetailMock,
      } as unknown as ReturnType<typeof service.getHeroDetail>)
    );

    const { asFragment } = render(
      <FavsProvider>
        {await HeroDetail({ params: { id: "1234-" } })}
      </FavsProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
