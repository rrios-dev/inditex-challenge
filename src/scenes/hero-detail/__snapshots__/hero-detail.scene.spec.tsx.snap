// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hero Detail Scene > Mount Scene with data 1`] = `
<DocumentFragment>
  <main
    class="_flex_6eb20e _dir-column_6eb20e _gap-12_6eb20e"
  >
    <section
      class="_hero-header-section_c43771 _flex_6eb20e"
    >
      <div
        class="_hero-header-section-content_c43771"
      >
        <img
          alt="Thumbnail image of the hero undefined"
          class="_hero-header-section-image_c43771"
          src="undefined.undefined"
        />
        <div
          class="_hero-header-section-info_c43771 _flex_6eb20e _dir-column_6eb20e _justify-center_6eb20e _gap-7_6eb20e"
        >
          <div
            class="_flex_6eb20e _justify-between_6eb20e _align-center_6eb20e _gap-4_6eb20e"
          >
            <h1
              class="_typography_142002 _variant-h1_142002 color-fg"
            />
            <button
              class="_button-base_31eab0"
            >
              <img
                alt="Empty heart"
                data-nimg="1"
                decoding="async"
                height="22"
                loading="lazy"
                src="/icons/heart-empty.svg"
                style="color: transparent;"
                width="24"
              />
            </button>
          </div>
        </div>
      </div>
    </section>
    <div
      class="_section_272693 _hero-comics-section_3581eb _flex_6eb20e _dir-column_6eb20e _gap-12_6eb20e"
    >
      <h2
        class="_typography_142002 _variant-h2_142002 color-fg _title_272693"
      >
        Comics
      </h2>
    </div>
  </main>
</DocumentFragment>
`;
